### PR TITLE
perf: Improve performance of k-d tree

### PR DESCRIPTION
**Pull request by stephenswat**
This pull request makes some major changes to the k-d tree in order to improve its performance. In particular, the following changes are made:

1. Internal nodes and leaf nodes are no longer virtual subtypes of an abstract base class, removing the need for v-table lookups during the use of the k-d tree.
2. Callables are now passed as universal references rather than copied `std::function` object, allowing the compiler to optimize more aggressively.
3. Some checks which should always hold are turned into assertions so they are not compiled in release builds.

These changes effect a 71% performance increase in the orthogonal seeding strategy in ITk examples, and (of course) the k-d tree tests pass.

Also gets rid of some unused methods and class members.

### CODE REVIEW

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 1]
```diff
@@ -318,369 +310,249 @@ class KDTree {
   /// A k-d tree consists of two different node types: leaf nodes and inner
   /// nodes. These nodes have some common functionality, which is captured by
   /// this common parent node type.
-  class KDTreeAbstractNode {
+  class KDTreeNode {
    public:
+    /// @brief Enumeration type for the possible node types (internal and leaf).
+    enum class NodeType { INTERNAL, LEAF };
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/KDTree.hpp:316]: Existing style would favor:
```suggestion
    enum class NodeType { Internal, Leaf };
```

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 2]
```diff
@@ -318,369 +310,249 @@ class KDTree {
...
-    /// @return The number of elements below this node.
-    std::size_t size(void) const { return std::distance(m_begin_it, m_end_it); }
-
-    /// @brief The axis-aligned bounding box containing all elements in this
-    /// node.
-    ///
-    /// @return The minimal axis-aligned bounding box that contains all the
-    /// elements under this node.
-    const range_t &range(void) const { return m_range; }
+        for (iterator_t i = this->m_begin_it; i != this->m_end_it; ++i) {
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/KDTree.hpp:348]: We usually omit `this`, but you were using this in the previous version of the code already, so it's fine.

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 3]
```diff
@@ -318,369 +310,249 @@ class KDTree {
...
+          for (std::size_t j = 0; j < Dims; ++j) {
+            // We normalize the values to be in the [0, 1] range, because
+            // otherwise the dimension with the higher valued elements would
+            // always win.
+            Scalar value = (i->first[j] - this->m_range[j].min()) /
+                           (this->m_range[j].max() - this->m_range[j].min());
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/KDTree.hpp:354]: Is this guaranteed to never DIVBYZERO?

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 4]
```diff
@@ -318,369 +310,249 @@ class KDTree {
...
+          }
+
+          return;
+        }
 
-      // Print some information about the node.
-      out << "Node with range " << this->m_range[m_dim].min() << " to "
-          << this->m_mid << " to " << this->m_range[m_dim].max()
-          << " (total size " << this->size() << ")" << std::endl;
+        assert(this->m_lhs && this->m_rhs);
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/KDTree.hpp:493]: Consider adding a message like
```suggestion
        assert(this->m_lhs && this->m_rhs && "Did not find lhs and rhs");
```

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 5]
```diff
@@ -318,369 +310,249 @@ class KDTree {
...
-        // In case this node's bounding box is fully contained within the
-        // range, we don't actually need to check this.
-        if (contained || r.contains(i->first)) {
-          f(i->first, i->second);
-        }
-      }
-    }
+    /// @return The minimal axis-aligned bounding box that contains all the
+    /// elements under this node.
+    const range_t &range(void) const { return m_range; }
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/KDTree.hpp:534]:
```suggestion
    const range_t &range() const { return m_range; }
```

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 6]
```diff
@@ -318,369 +310,249 @@ class KDTree {
...
-      // There isn't much to say about leaf nodes...
-      out << "Leaf with " << this->size() << " objects" << std::endl;
+    /// @brief The axis-aligned bounding box of the coordinates under this
+    /// node.
+    const range_t m_range;
 
-      return out.str();
-    }
+    /// @brief Pointers to the left and right children.
+    std::unique_ptr<KDTreeNode> m_lhs, m_rhs;
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/KDTree.hpp:548]:
```suggestion
    std::unique_ptr<KDTreeNode> m_lhs
    std::unique_ptr<KDTreeNode> m_rhs;
```

**Review by paulgessinger — COMMENTED** [review]
Comments are largely stylistic.

I'm ok to merge this without changing `(void)` and `this`, but I'd encourage you to make a follow-up PR to bring this in line with the rest of the code base.

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 1 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/KDTree.hpp:316]: Fixed.

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 4 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/KDTree.hpp:493]: Done.

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 7]
```diff
@@ -318,369 +310,249 @@ class KDTree {
...
-    /// doesn't do.
+    /// Conveniently, this number is always equal to the distance between the
+    /// begin iterator and the end iterator, so we can simply delegate to the
+    /// relevant standard library method.
     ///
-    /// @param _b The begin iterator of the range of values.
-    /// @param _e The end iterator of the range of values.
-    KDTreeLeaf(iterator_t _b, iterator_t _e) : KDTreeAbstractNode(_b, _e) {}
+    /// @return The number of elements below this node.
+    std::size_t size(void) const { return std::distance(m_begin_it, m_end_it); }
```
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/KDTree.hpp:527]: Removed.

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 5 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/KDTree.hpp:534]: Removed.

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 6 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/KDTree.hpp:548]: Split.

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 3 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/KDTree.hpp:354]: I took the liberty of removing this code now, but you're right; this was not necessarily safe.

**Comment by asalzburger** [comment]
Ok, I rebased, would be nice to get this in and provide it to Luis for testing.

**Comment by stephenswat** [comment]
Marking WIP because I plan to make a few more modifications.

**Comment by asalzburger** [comment]
Hey @stephenswat - will you pick up on that?

**Comment by stephenswat** [comment]
This PR is now ready for review.

**Review by asalzburger — APPROVED** [review]
Good to go from my side.

**Comment by LuisFelipeCoelho** [comment]
The physmon fail is a small change in physics performance, but I think we can just replace the reference root file

**Comment by stephenswat** [comment]
The changes in physmon make sense; the seeding is sensitive to the order in which spacepoints are received, which this pull request can change. I will update the references and we can go ahead.

**Review by paulgessinger — APPROVED** [review]

**Comment by stephenswat** [comment]
Ready to go, but needs re-approval.

**Comment by stephenswat** [comment]
Wow, telepathy.

**Pull request merged**